src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -164,103 +157,83 @@ public boolean partitionExists(final String topic, final String partition) throw
+     *
+     * It allows to not send messages to Kafka when circuit breaker is opened.
+     *
+     * @see org.zalando.nakadi.repository.kafka.ProducerSendCommand
+     * @see org.zalando.nakadi.repository.kafka.ProducerSentCallbackCommand
+     *
+     * @param topicId
+     * @param batchItems
+     * @throws EventPublishingException
+     */
     @Override
-    public void syncPostBatch(final String topicId, final List<BatchItem> batch) throws EventPublishingException {
-        batch.forEach(item -> Preconditions.checkNotNull(
-                item.getPartition(), "BatchItem partition can't be null at the moment of publishing!"));
+    public void syncPostBatch(final String topicId, final List<BatchItem> batchItems) throws EventPublishingException {
         final Producer<String, String> producer = kafkaFactory.takeProducer();
-        final Map<BatchItem, CompletableFuture<Exception>> sendFutures = new HashMap<>();
         try {
-            for (final BatchItem item : batch) {
-                item.setStep(EventPublishingStep.PUBLISHING);
-                sendFutures.put(item, publishItem(producer, topicId, item));
-            }
-            final CompletableFuture<Void> multiFuture = CompletableFuture.allOf(
-                    sendFutures.values().toArray(new CompletableFuture<?>[sendFutures.size()]));
-            multiFuture.get(createSendTimeout(), TimeUnit.MILLISECONDS);
-
-            // Now lets check for errors
-            final Optional<Exception> needReset = sendFutures.entrySet().stream()
-                    .filter(entry -> isExceptionShouldLeadToReset(entry.getValue().getNow(null)))
-                    .map(entry -> entry.getValue().getNow(null))
-                    .findAny();
-            if (needReset.isPresent()) {
-                LOG.info("Terminating producer while publishing to topic {} because of unrecoverable exception",
-                        topicId, needReset.get());
-                kafkaFactory.terminateProducer(producer);
+            final Map<String, String> partitionToBroker = producer.partitionsFor(topicId).stream()
+                    .collect(Collectors.toMap(p -> String.valueOf(p.partition()),p -> String.valueOf(p.leader().id())));
+            batchItems.forEach(item -> {
+                Preconditions.checkNotNull(
+                        item.getPartition(), "BatchItem partition can't be null at the moment of publishing!");
+                item.setBrokerId(partitionToBroker.get(item.getPartition()));
+                item.setTopic(topicId);
+            });
+
+            // send message to Kafka in asynch mode, it will fail sending in circuit is opened
+            final List<ProducerSendCommand.NakadiCallback> callbacks = batchItems.stream()
+                    .peek(item -> item.setStep(EventPublishingStep.PUBLISHING))
+                    .map(item -> new ProducerSendCommand(producer, item).execute())
```

Comment:
@adyach I know this might be too late feedback. I'm afraid that creating one `ProducerSendCommand` per item might cause short circuit too often and too early. We have some batches with hundreds of events. If one batch times out, we would have an open circuit caused by a single request.

If that's the case, is it possible for us to disable Hystrix? I mean, if my hypothesis is proven to be true and we have open circuits when we don't want it to open, is it possible to disable hystrix?